src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc

```diff
@@ -16,41 +16,261 @@
  *
  */
 
+#include <grpc/support/port_platform.h>
+
 #include "src/core/tsi/s2a/handshaker/s2a_handshaker_client.h"
 
+#include <grpc/support/alloc.h>
+#include <grpc/support/log.h>
+
+#include "src/core/lib/slice/slice_internal.h"
+#include "src/core/lib/surface/call.h"
+#include "src/core/lib/surface/channel.h"
+#include "src/core/tsi/s2a/handshaker/s2a_handshaker_util.h"
+#include "src/core/tsi/s2a/s2a_constants.h"
+#include "src/proto/grpc/gcp/s2a.upb.h"
+
 namespace grpc_core {
 namespace experimental {
 
-tsi_result s2a_handshaker_client_start_client(
-    const s2a_handshaker_client* client) {
-  return TSI_UNIMPLEMENTED;
+/** ------------ Preparation of client start messages. --------------- **/
+
+grpc_byte_buffer* s2a_handshaker_client::s2a_get_serialized_start_client() {
+  upb::Arena arena;
+  s2a_SessionReq* request = s2a_SessionReq_new(arena.ptr());
+  s2a_ClientSessionStartReq* start_client =
+      s2a_SessionReq_mutable_client_start(request, arena.ptr());
+  s2a_ClientSessionStartReq_add_application_protocols(
+      start_client, upb_strview_makez(kS2AApplicationProtocol), arena.ptr());
+  int32_t* tls_versions = s2a_ClientSessionStartReq_resize_tls_versions(
+      start_client, /*len=*/1, arena.ptr());
+  GPR_ASSERT(tls_versions != nullptr);
+  tls_versions[0] = s2a_TLS1_3;
+  int32_t* tls_ciphersuites = s2a_ClientSessionStartReq_resize_tls_ciphersuites(
+      start_client, options_->supported_ciphersuites().size(), arena.ptr());
+  GPR_ASSERT(tls_ciphersuites != nullptr);
+  for (size_t i = 0; i < options_->supported_ciphersuites().size(); i++) {
+    tls_ciphersuites[i] = options_->supported_ciphersuites()[i];
+  }
+  for (size_t j = 0; j < options_->target_service_account_list().size(); j++) {
+    s2a_Identity* identity = s2a_ClientSessionStartReq_add_target_identities(
+        start_client, arena.ptr());
+    s2a_Identity_set_spiffe_id(
+        identity,
+        upb_strview_makez(options_->target_service_account_list()[j].c_str()));
+  }
+  s2a_ClientSessionStartReq_set_target_name(
+      start_client, upb_strview_make(reinterpret_cast<const char*>(
+                                         GRPC_SLICE_START_PTR(target_name_)),
+                                     GRPC_SLICE_LENGTH(target_name_)));
+  return s2a_get_serialized_session_req(request, arena.ptr());
+}
+
+tsi_result s2a_handshaker_client::client_start() {
+  grpc_byte_buffer* buffer = s2a_get_serialized_start_client();
+  if (buffer == nullptr) {
+    gpr_log(GPR_ERROR, kS2AGetSerializedStartClientFailed);
+    return TSI_INTERNAL_ERROR;
+  }
+  grpc_byte_buffer_destroy(send_buffer_);
+  send_buffer_ = buffer;
+  tsi_result call_result = make_grpc_call(/*is_start=*/true);
+  if (call_result != TSI_OK) {
+    gpr_log(GPR_ERROR, kS2AMakeGrpcCallFailed);
+  }
+  return call_result;
+}
+
+/** ------------ Preparation of server start messages. --------------- **/
+
+grpc_byte_buffer* s2a_handshaker_client::s2a_get_serialized_start_server(
```

Comment:
Please add comments within this function e.g., what parts of handshake messages are being set.

Reply:
Same for other API in this file that prepares the handshake messages.

Reply:
I've added comments to the declarations of `StartClient`, `StartServer`, and `Next` in `s2a_handshaker_client.cc`.